### ISSUE TITLE
Subqueries in the target list isn't supported

### ISSUE THREAD

**Issue by ot-citysurf** [citusdata/citus#2010]
Error msg: "Analysis failed: could not run distributed query with subquery outside the FROM and WHERE clauses"

We are using PostgreSQL 10.2 and Citus 7.2-3 on a Linux 16 platform. Since we are aiming to do spatial analysis with a huge spatial data, we have installed the last commit of crankshaft extension developed by CartoDB team for PostgreSQL. This is a standalone extension and only requires postgis and plpythonu installations. We are planning to use Citus Data as a back-end platform for our analysis tool & service which allows multiple user access at a time and query latency is very crucial for us.

When we use reference tables, the queries for every function call are executed successfully. However, when we use distributed tables, we are getting the specified error above for each function. Here is an example query:

SELECT m.*
FROM cs_glb.CDB_GetisOrdsG('SELECT * FROM cs_glb.istanbul_hasta',                              'alt_tani_grubu','',5,99,'the_geom','id') As m

The subquery provides dataset for the analysis.

**Comment by ozgune** [comment]
@onderkalaci or @metdos -- Could you check if we provide support for this query type?

**Comment by onderkalaci** [comment]
### Short answer:
Citus supports this type of subqueries in general. However, `CDB_GetisOrdsG ()` function gets the query and converts into the following form, which is not supported (Given that it contains subqueries in the target list i.e., the subqueries exists in the `SELECT` clause, not `FROM` or `WHERE`)

```SQL
SELECT id,
       (alt_tani_grubu)::numeric AS attr1,
  (SELECT ARRAY
     (SELECT j.id
      FROM
        (SELECT yil_ay, hasta_ili, hasta_ilcesi, muayene_ili, muayene_ilce, muayene_kurum, paydas, kurum_turu, turkiye_nufus, il_nufus, ilce_nufus, ust_tani_grubu, alt_tani_grubu, brans, yatis_mi, nuts3_bolge_kodu, nuts3_bolge_adi, nuts3_kurum_bolge_kodu, nuts3_kurum_bolge_adi, toplam_muayene_sayisi, csbmkodu, tekil_hasta_sayisi, the_geom, id
         FROM cdb_crankshaft.istanbul_hasta_3000) j
      WHERE ((i.id <> j.id)
             AND public.st_touches(i.the_geom, j.the_geom)
             AND (j.alt_tani_grubu IS NOT NULL))) AS "array") AS neighbors
FROM
  (SELECT yil_ay,
          hasta_ili,
          hasta_ilcesi,
          muayene_ili,
          muayene_ilce,
          muayene_kurum,
          paydas,
          kurum_turu,
          turkiye_nufus,
          il_nufus,
          ilce_nufus,
          ust_tani_grubu,
          alt_tani_grubu,
          brans,
          yatis_mi,
          nuts3_bolge_kodu,
          nuts3_bolge_adi,
          nuts3_kurum_bolge_kodu,
          nuts3_kurum_bolge_adi,
          toplam_muayene_sayisi,
          csbmkodu,
          tekil_hasta_sayisi,
          the_geom,
          id
   FROM cdb_crankshaft.istanbul_hasta_3000) i
WHERE (alt_tani_grubu IS NOT NULL)
ORDER BY id
```

### Workaround: 

Citus introduced a new planning logic with #1804. We're still expanding it, and this type of queries (where subqueries in the target list and similarly subqueries in HAVING clause) is in our scope in the future.

Today, however, we have a trick to enable recursive planning  **in some cases** by making the subqueries not pushdownable. In other words, if you add an `OFFSET 0` clause to the subqueries, which doesn't change the query result at all, the query is forced to be planned by Citus. So, for this example, the following query becomes supported:

```SQL
SELECT 
    m.*
FROM 
    cs_glb.CDB_GetisOrdsG('SELECT * FROM cs_glb.istanbul_hasta OFFSET 0', 'alt_tani_grubu','',5,99,'the_geom','id') As m
```

**Comment by ot-citysurf** [comment]
Thank you for your detailed examination and support. This trick solution works for all of our queries.

**Comment by onderkalaci** [comment]
Another way to reproduce this:
```SQL

CREATE TABLE users (id text PRIMARY KEY, 
                    name text, 
                    created_at timestamptz DEFAULT now(), 
                    last_see  timestamptz);

CREATE TABLE todos (id bigserial PRIMARY KEY, 
                    title text, 
                    is_completed boolean, 
                    is_public boolean, 
                    created_at timestamp DEFAULT now(), 
                    user_id text);

-- these above restrictions brought us to the following schema
SELECT create_reference_table('users');
SELECT create_distributed_table('todos', 'id');

alter table "public"."todos"
                   add constraint "todos_user_id_fkey"
                   foreign key ("user_id")
                   references "public"."users"
                   ("id") on update restrict on delete restrict;


SELECT coalesce(json_agg(root), '[]') AS root
FROM
  (SELECT row_to_json(
                        (SELECT _1_e
                         FROM
                           (SELECT "_0_root.base".id AS id, 
                                   "_0_root.base".title AS title, 
                                   "_0_root.base".is_completed AS is_completed, 
                                   "_0_root.base".is_public AS is_public, 
                                   "_0_root.base".created_at AS created_at, 
                                   "_0_root.base".user_id AS user_id) AS _1_e)) AS root
   FROM
     (SELECT *
      FROM public.todos) AS "_0_root.base"
   LIMIT 10
   OFFSET 0) AS _2_root;
ERROR:  could not run distributed query with subquery outside the FROM, WHERE and HAVING clauses
HINT:  Consider using an equality filter on the distributed table's partition column.
```

**Comment by onderkalaci** [comment]
I think this issue is resolved with #4360
